besmarts-core/python/examples/07_besmarts_fit.py

```diff
@@ -100,54 +127,70 @@ def make():
     return d
 
 
-def new_gdb(f: Dict[str, List[str]]) -> assignments.graph_db:
+def new_gdb(f: Dict[str, List[Dict[int, str]]]) -> assignments.graph_db:
+    """
+    This creates a ``graph_db`` from a dataset
+    as generated by ``make()``.
+    """
     gcd = codec_rdkit.graph_codec_rdkit()
     gdb = assignments.graph_db()
 
-    for smi, fn_dict in f.items():
+    for smi, fn_list_of_dicts in f.items():
+        # create a graph and add it to the graph_db
+        g: graphs.graph = gcd.smiles_decode(smi)
+        # molecule_graph_id is the graph ID of the graph in the graph_db
+        molecule_graph_id: int = assignments.graph_db_add_graph(gdb, smi, g)
```

Comment:
Is it possible to use besmarts with multi-molecule data (e.g. dimers) at all?

Reply:
Stuff like dimers is the reason I made the much more complicated graph_db data structure.. however right now nearly everything assumes a single molecule, so it would need some work/checking to get going correctly. Each molecule coordinates would be a graph_db_column and the graph_db_row could be considered a single snapshot/frame, where each of these would be associated to the graph_db_graph of the same graph ID.

Reply:
Ahh I *think* I understand. So can a `graph.graph` contain multiple molecules? IIRC the graph_same check only checks encoded primitives so coordinates wouldn't differentiate the molecules. So I'm wondering if a dimer energy of identical molecules would need to reference the same graph ID twice, and how that would interact with `graph_db_tables.gid`.

(Just to check my understanding -- does the below sound right to you?)
Each graph_db_entry in the gdb contains graph_db_tables, which contains graph_db_graphs that are referenced by `gid`. Coordinates are stored separately by atom in graph_db_columns, which are collected in graph_db_rows. In the case of using an interaction energy between two different molecules, I think the value at `gde.tables[assignments.ENERGIES].values[snapshot_i]` would be the interaction energy at `gde.tables[assignments.POSITIONS].graphs[gid].rows[snapshot_i]`?

Reply:
I've been treating all graphs as fully connected; I haven't gone into much thinking/testing whether disconnected graphs should be allowed as of yet. There are pros and cons to doing so, but right now I'm leaning towards having everything connected (and maybe adding empty bonds for specific problems; these would encode to . in the string representation). I believe everything in `graphs` is strictly 2D based, so yes the graph operations won't pick up on different conformations. The dimer would most certainly reference the same graph ID; the underlying graph is how selections are referenced in the graph_db_columns (if you have IC data; perhaps less relevant here). As an example you can query bonds of a graph, say (1, 2) by calling `graphs.graph_bonds(gdb.graphs[gid])`, which would be a selection in the graph_db_column, such as `gde.tables[BONDS][gid][rid][cid][(1,2)]` (if you put data there).  There is no way to specify conformations to a graph object; think of them as part of a regular FF topology. The easiest way to do so would be a `assignments.graph_assignment` object, which are used in most places needing positions.

And yes, that is a quick way to make that work and probably the only sane way to do so, showing the usefulness of having a dumb values member :) I read that as the total energy for a particular snapshot. There may be two gids each with one column for `snapshot_i`, or the same gid with two columns with different sets of coordinates.